runtime/src/iree/hal/drivers/hip/util/queue.c

```diff
@@ -0,0 +1,90 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/hal/drivers/hip/util/queue.h"
+
+#include "iree/base/api.h"
+
+void iree_hal_hip_util_queue_initialize(iree_allocator_t allocator,
+                                        iree_host_size_t element_size,
+                                        iree_host_size_t inline_count,
+                                        iree_hal_hip_util_queue_t* out_queue) {
+  out_queue->allocator = allocator;
+  out_queue->elements = &out_queue->initial_allocation[0];
+  out_queue->element_size = element_size;
+  out_queue->element_count = 0;
+  out_queue->capacity = inline_count;
+  out_queue->head = 0;
+}
+
+void iree_hal_hip_util_queue_deinitialize(iree_hal_hip_util_queue_t* queue) {
+  IREE_ASSERT_ARGUMENT(queue);
+  if (queue->elements != &queue->initial_allocation[0]) {
+    iree_allocator_free(queue->allocator, queue->elements);
+  }
+}
+
+iree_status_t iree_hal_hip_util_queue_push_back(
+    iree_hal_hip_util_queue_t* queue, void* element) {
+  // Expand the queue if necessary.
+  if (queue->capacity == queue->element_count) {
+    uint8_t* new_mem = NULL;
+    queue->capacity = iree_max(16, queue->capacity << 1);
```

Comment:
readability: prefer arithmetic unless doing bit math or something that's explicitly requiring the bit math (compilers will make this << if they can)
```suggestion
    queue->capacity = iree_max(16, queue->capacity * 2);
```